tests/api_tests.py

```diff
@@ -131,15 +133,127 @@ async def test_immunization_happy_path(api_client: APISessionClient, authorised_
 
 @pytest.mark.e2e
 @pytest.mark.asyncio
-async def test_bad_nhs_number(api_client: APISessionClient, authorised_headers):
+@pytest.mark.parametrize(
+    "request_data",
+    [
+        # condition 1: invalid iss claim
+        {
+            "expected_status_code": 401,
+            "expected_response": {
+                "severity": "error",
+                "error_code": "processing",
+                "error_diagnostics": "Missing or invalid 'iss' claim in ID Token",
+            },
+            "claims": {
+                "iss": "invalid"
+            }
+        },
+        # condition 2: invalid typ header
+        {
+            "expected_status_code": 401,
+            "expected_response": {
+                "severity": "error",
+                "error_code": "processing",
+                "error_diagnostics": "Missing or invalid 'typ' header in ID Token - must be 'JWT'",
+            },
+            "headers": {
+                "typ": "invalid"
+            }
+        },
+        # condition 3: invalid identity_proofing_level claim
+        {
+            "expected_status_code": 401,
+            "expected_response": {
+                "severity": "error",
+                "error_code": "processing",
+                "error_diagnostics": "Missing or invalid 'identity_proofing_level' claim in ID Token",
+            },
+            "claims": {
+                "identity_proofing_level": "invalid"
+            }
+        },
+        # condition 4: jwt expired
+        {
+            "expected_status_code": 401,
+            "expected_response": {
+                "severity": "error",
+                "error_code": "processing",
+                "error_diagnostics": "Invalid exp claim in JWT - JWT has expired",
+            },
+            "claims": {
+                "exp": int(time()) - 10,  # Set JWT as already expired
+                "iat": int(time()) - 10
+            }
+        },
+        # condition 5: invalid id token
+        {
+            "expected_status_code": 401,
+            "expected_response": {
+                "severity": "error",
+                "error_code": "processing",
+                "error_diagnostics": "Malformed JWT in 'NHSD-User-Identity' header",
+            },
+            "id_token": "invalid"
+        },
+    ],
+)
+async def test_immunisation_id_token_error_scenarios(test_app,
+                                                     api_client: APISessionClient,
+                                                     authorised_headers, request_data: dict):
+    sleep(1)  # Add delay to tests to avoid 429 on service callout
```

Comment:
I think 
```await asyncio.sleep(1)```
is better here, it yiellds the main thread back to other tests for that 1 second.

Reply:
Good shout. Adding

Reply:
Ticket for refactor in future as the async method didn't work for these tests - https://nhsd-jira.digital.nhs.uk/browse/APM-2031